types/wlr_drm_lease_v1.c

```diff
@@ -0,0 +1,645 @@
+	if (fd < 0) {
+		wlr_log(WLR_ERROR, "Unable to get read only DRM fd for leasing");
+		wl_client_post_no_memory(wl_client);
+		return;
+	}
+
+	wp_drm_lease_device_v1_send_drm_fd(lease_resource, fd);
+	close(fd);
+
+	wl_list_insert(&device->resources, wl_resource_get_link(lease_resource));
+
+	struct wlr_drm_lease_connector_v1 *connector;
+	wl_list_for_each(connector, &device->connectors, link) {
+		drm_lease_connector_v1_send_to_client(connector, lease_resource);
+	}
+}
+
+static void handle_output_destroy(struct wl_listener *listener, void *data) {
+	struct wlr_drm_lease_connector_v1 *conn = wl_container_of(listener, conn,
+			destroy);
+
+	wlr_log(WLR_DEBUG, "Handle destruction of output  %s", conn->output->name);
+	if (conn->active_lease) {
+		wlr_drm_lease_v1_revoke(conn->active_lease);
+	}
+	wlr_drm_lease_v1_manager_withdraw_output(conn->device->manager, conn->output);
+
+	drm_lease_connector_v1_destroy(conn);
+}
+
+bool wlr_drm_lease_v1_manager_offer_output(
+		struct wlr_drm_lease_v1_manager *manager, struct wlr_output *output) {
+	assert(manager && output);
+	assert(wlr_output_is_drm(output));
+
+	wlr_log(WLR_DEBUG, "Offering output %s", output->name);
+
+	struct wlr_drm_lease_device_v1 *device = NULL, *tmp_device;
+	wl_list_for_each(tmp_device, &manager->devices, link) {
+		if (tmp_device->backend == output->backend) {
+			device = tmp_device;
+			break;
+		}
+	}
+
+	if (!device) {
+		wlr_log(WLR_ERROR, "No wlr_drm_lease_device_v1 associated with the "
+				"offered output");
+		return false;
+	}
+
+	struct wlr_drm_lease_connector_v1 *connector;
+	wl_list_for_each(connector, &device->connectors, link) {
+		if (connector->output == output) {
+			wlr_log(WLR_ERROR, "Output %s has already been offered",
+					output->name);
+			return false;
+		}
+	}
+
+	connector = calloc(1, sizeof(struct wlr_drm_lease_connector_v1));
+	if (!connector) {
+		wlr_log(WLR_ERROR, "Failed to allocatr wlr_drm_lease_connector_v1");
+		return false;
+	}
+
+	connector->output = output;
+	connector->device = device;
+
+	wl_signal_add(&output->events.destroy, &connector->destroy);
+	connector->destroy.notify = handle_output_destroy;
+
+	wl_list_init(&connector->resources);
+	wl_list_insert(&device->connectors, &connector->link);
+
+	struct wl_resource *resource;
+	wl_resource_for_each(resource, &device->resources) {
+		drm_lease_connector_v1_send_to_client(connector, resource);
+	}
+
+	return true;
+}
+
+
+void wlr_drm_lease_v1_manager_withdraw_output(
+	struct wlr_drm_lease_v1_manager *manager, struct wlr_output *output) {
+	assert(manager && output);
+
+	wlr_log(WLR_DEBUG, "Withdrawing output %s", output->name);
+
+	struct wlr_drm_connector *drm_connector =
+			get_drm_connector_from_output(output);
+
+	struct wlr_drm_lease_device_v1 *device = NULL, *tmp_device;
+	wl_list_for_each(tmp_device, &manager->devices, link) {
+		struct wlr_drm_backend *backend = get_drm_backend_from_backend(
+				tmp_device->backend);
```

Comment:
No need to compare `struct wlr_drm_backend *` values here, we can compare `struct wlr_backend *` structs instead.

```c
if (tmp_device->backend == output->backend) {
```